proposals/lip-0045.md

```diff
@@ -779,199 +743,193 @@ The `apply` function applies a CCM. It is specified in [LIP 0049][lip-0049].
-
-##### Execution
+The `bounce` and `bounceWithoutFee` functions return a CCM that errored during the CCM execution stages to the sending chain. 
+`bounceWithoutFee` is called if an error occurs during the cross-chain command execution (see [`apply` function][lip-0049#apply]) in the receiving chain. 
+In this case, the fee property is set to 0 and the message is returned only if the fee was larger than a minimum return fee and the status of the CCM is `CCM_OK`.
+`bounce` is called if an error occurs in any other stage.
+In this case, the fee is not changed and the message is returned if the status of the CCM is `CCM_OK`.
 
 ```python
-sendInternal(feeAddress, moduleID, crossChainCommandID, receivingChainID, fee, status, params):
-    # Processing on the mainchain
-    if ownChainAccount.ID == MAINCHAIN_ID:
-        # Chain has to exist on mainchain
-        if chainAccount(receivingChainID) does not exists:
-            return False
-        partnerChainID = receivingChainID
-    # Processing on a sidechain
-    else:
-        # Check for direct channel
-        if chainAccount(receivingChainID) does not exists:
-            partnerChainID = mainchainID
-        else:
-            partnerChainID = receivingChainID
-
-    # Chain must be live; This checks is always on the receivingChainID
-    if not isLive(receivingChainID):
-        return False
-    # Chain status must be active
-    if not chainAccount(partnerChainID).status == CHAIN_ACTIVE:
-        return False
-
-    # Create cross-chain message
-    ccm = {
-        "nonce": ownChainAccount.nonce,
-        "moduleID": moduleID,
-        "crossChainCommandID": crossChainCommandID,
-        "sendingChainID": ownChainAccount.ID,
-        "receivingChainID": receivingChainID,
-        "fee": fee,
-        "status": status,
-        "params": params
-    }
-
-    if validateFormat(ccm) == False:
-        return False
-
-    for each module mdl for which beforeSendCCM exists:
-        mdl.beforeSendCCM(feeAddress, ccm)
-        if the previous call fails:
-            return False
-
-    addToOutbox(partnerChainID, ccm)
-    ownChainAccount.nonce += 1
+def bounce(ccm: CCM, errorCode: uint32) -> None:
+    ccmID = sha256(encode(crossChainMessageSchema, ccm))
+    emitEvent(
+        module = MODULE_NAME_INTEROPERABILITY,
+        name = EVENT_NAME_CCM_PROCESSED,
+        data = {"ccmID": ccmID, "result": errorCode},
+        topics = [ccm.sendingChainID, ccm.receivingChainID]
+    )
 
-    return True
+    if ccm.status == CCM_OK:
+        ccm.status = errorCode
+        ccm.sendingChainID, ccm.receivingChainID = ccm.receivingChainID, ccm.sendingChainID
+        addToOutbox(ccm.receivingChainID, ccm)
+        newCcmID = sha256(encode(crossChainMessageSchema, ccm))
+        emitEvent(
+            module = MODULE_NAME_INTEROPERABILITY,
+            name = EVENT_NAME_CCM_PROCESSED,
+            data = {"ccmID": newCcmID, "result": CCM_BOUNCED},
```

Comment:
What is confusing here is that usually the CCM Processed event has the ID of the CCM being processed as data where instead here it is a new CCM being emitted.

Reply:
Wouldn't it be better to use the `EVENT_NAME_CCM_SENT` event?

Reply:
done